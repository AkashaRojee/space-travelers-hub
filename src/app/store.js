import { configureStore } from '@reduxjs/toolkit';
import rocketsReducer from '../features/rockets/rocketSlice';

export default configureStore({
  reducer: {
    rockets: rocketsReducer,
  },
});
