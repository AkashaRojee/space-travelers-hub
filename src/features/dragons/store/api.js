// ACTION TYPES
const API_REQUESTED = 'api/requested';
const API_CALL_SUCCESS = 'api/success';
const API_CALL_FAIL = 'api/fail';

// ACTION CREATORS

export const requestAPICall = (payload) => ({
  type: API_REQUESTED,
  payload,
});
export const onAPICallSuccess = () => ({
  type: API_CALL_SUCCESS,
});
export const onAPICallFail = () => ({
  type: API_CALL_FAIL,
});
