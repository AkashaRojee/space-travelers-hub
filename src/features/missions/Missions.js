import { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { getMissions } from './missionsSlice';
const Missions = () => {
	const dispatch = useDispatch();
	const { missions } = useSelector((state) => state).missions;
	console.log(missions);
	useEffect(() => {
		dispatch(getMissions());
	}, [dispatch]);
	return (
		<section className='missions-section'>
			<table className="mission-table">
				<thead>
					<tr>
						<th>Missions</th>
						<th>Descripition</th>
						<th>Status</th>
						<th></th>
					</tr>
				</thead>
				<tbody>
          <tr>
            <td>Alfreds Futterkiste</td>
            <td>
              Lorem ipsum dolor sit amet consectetur adipisicing elit. Delectus
              porro necessitatibus maiores sequi. Dicta libero facere iste
              adipisci odit recusandae quae fuga impedit cupiditate, sint deleniti
              laboriosam
            </td>
            <td>
              <span>Not a member</span>
            </td>
            <td>
              <button>Join Mission</button>
            </td>
          </tr>
        </tbody>
			</table>
		</section>
	);
};

export default Missions;
