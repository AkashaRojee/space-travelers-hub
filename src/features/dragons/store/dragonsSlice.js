import * as apiActions from './api';

// ACTION TYPES
const DRAGONS_REQUESTED = 'dragons/request';
const DRAGONS_LOADED = 'dragons/load';
const DRAGON_RESERVED = 'dragon/:id/reserve';
const DRAGON_UNRESERVED = 'dragon/:id/unreserve';

// ACTION CREATORS
export const loadDragons = () => (dispatch, getState) => {
  const {
    dragons: { lastFetched },
  } = getState();
  return lastFetched
    ? getState()
    : dispatch(
        apiActions.requestAPICall({
          url: 'https://api.spacexdata.com/v3/dragons',
          onStart: DRAGONS_REQUESTED,
          onSuccess: DRAGONS_LOADED,
        })
      );
};
export const reserveDragon = (id) => ({
  type: DRAGON_RESERVED,
  payload: { id },
});
export const unReserveDragon = (id) => ({
  type: DRAGON_UNRESERVED,
  payload: { id },
});

// REDUCER
const initialState = {
  list: [],
  isLoading: false,
  lastFetched: null,
};
const reducer = (state = initialState, action) => {
  const { type, payload } = action;
  if (type === DRAGONS_REQUESTED) return { ...state, isLoading: true };
  if (type === DRAGONS_LOADED) {
    return {
      ...state,
      lastFetched: Date.now(),
      isLoading: false,
      list: payload.map(({ id, name, type, flickr_images: [imgSrc] }) => ({
        id,
        name,
        type,
        imgSrc,
        isReserved: false,
      })),
    };
  }
  if (type === DRAGON_RESERVED) {
    return {
      ...state,
      list: state.list.map((dragon) =>
        dragon.id === payload.id ? { ...dragon, isReserved: true } : dragon
      ),
    };
  }
  if (type === DRAGON_UNRESERVED) {
    return {
      ...state,
      list: state.list.map((dragon) =>
        dragon.id === payload.id ? { ...dragon, isReserved: false } : dragon
      ),
    };
  }
  return state;
};

export default reducer;
