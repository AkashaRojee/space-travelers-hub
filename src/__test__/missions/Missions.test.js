import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import Missions from '../../features/missions/Missions';
import store from '../../app/store';

const MockMissions = () => (
  <Provider store={store}>
    <Missions />
  </Provider>
);

describe('Mission component', () => {
  it('should match the snapshot', () => {
    const tree = renderer.create(<MockMissions />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
