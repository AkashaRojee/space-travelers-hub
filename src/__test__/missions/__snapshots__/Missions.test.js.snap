// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Mission component should match the snapshot 1`] = `
<section
  className="missions-section"
>
  <table
    className="mission-table"
  >
    <thead>
      <tr>
        <th>
          Missions
        </th>
        <th>
          Descripition
        </th>
        <th>
          Status
        </th>
        <th
          aria-label="Leave/Join Mission"
        />
      </tr>
    </thead>
    <tbody />
  </table>
</section>
`;
